semgrep-core/src/tainting/Dataflow_tainting.ml

```diff
@@ -383,128 +412,171 @@ let handle_taint_propagators env x taints =
   in
-  match sanitizer_pms with
+  let lval_env = Lval_env.add lval taints_sources_mut env.lval_env in
+  (Taints.union taints_sources_reg taints_sources_mut, lval_env)
+
+let rec check_tainted_lval env (lval : IL.lval) : Taints.t * Lval_env.t =
+  let new_taints, lval_in_env, lval_env = check_tainted_lval_aux env lval in
+  let taints_from_env = status_to_taints lval_in_env in
+  let taints = Taints.union new_taints taints_from_env in
+  let sinks = lval_is_sink env.config lval |> Common.map trace_of_match in
+  let findings = findings_of_tainted_sinks { env with lval_env } taints sinks in
+  report_findings { env with lval_env } findings;
+  (taints, lval_env)
+
+and check_tainted_lval_aux env (lval : IL.lval) :
+    Taints.t * [ `Clean | `None | `Tainted of Taints.t ] * Lval_env.t =
+  (* Recursively checks an l-value bottom-up.
+   *
+   *  This check needs to combine matches from pattern-{sources,sanitizers,sinks}
+   *  with the info we have stored in `env.lval_env`. This can be subtle, see
+   *  comments below.
+   *)
+  match lval_is_sanitized env.config lval with
   (* TODO: We should check that taint and sanitizer(s) are unifiable. *)
-  | _ :: _ ->
-      let lval_env' = sanitize_var env.lval_env sanitizer_pms var in
-      (Taints.empty, lval_env')
+  | _ :: _ as sanitizer_pms ->
+      (* NOTE [lval/clean]:
+       *  If lval is sanitized, then we will "bubble up" the `Clean status, so any
+       *  taint recorded in lval_env for any extension of lval will be discarded.
+       *
+       *  So, if we are checking `x.a.b.c` and `x.a` is clean (could be due to
+       *  sanitization) then any extension of `x.a` is considered clean as well,
+       *  and we do look for taint info in the environment.
+       *
+       *  *IF* sanitization is side-effectful then any taint info will be removed
+       *  from lval_env by sanitize_lval, but that is not guaranteed.
+       *)
+      let lval_env = sanitize_lval env.lval_env sanitizer_pms lval in
+      (Taints.empty, `Clean, lval_env)
   | [] ->
-      let mut_source_pms, reg_source_pms =
-        (* If the variable is an exact match (overlap > 0.99) for a source
-         * annotation, then we infer that the variable itself is now tainted
-         * (presumably by side-effect) and we will update the `var_env`
-         * accordingly. Otherwise the variable belongs to a piece of code that
-         * is a source of taint, but it is not tainted on its own. *)
-        List.partition (fun src -> src.overlap > 0.99) source_pms
+      (* Recursive call, check sub-lvalues first.
+       *
+       * It needs to be done bottom-up because any sub-lvalue can be a source and a
+       * sink by itself, even if an extension of lval is not. For example, given
+       * `x.a.b`, this lvalue may be considered sanitized, but at the same time `x.a`
+       * could be tainted and considered a sink in some context. We cannot just check
+       * `x.a.b` and forget about the sub-lvalues.
```

Comment:
Wait---is `x.a` a sub-lvalue of `x.a.b`? This is just a terminology thing.

Reply:
Yep, is it confusing ?